flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/SubtaskGatewayImpl.java

```diff
@@ -220,12 +228,11 @@ void openGatewayAndUnmarkCheckpoint(long checkpointId) {
 
         // Gateways should always be marked and closed for a specific checkpoint before it can be
         // reopened for that checkpoint. If a gateway is to be opened for an unforeseen checkpoint,
-        // exceptions should be thrown.
+        // which might happen when the coordinator has been reset to a previous checkpoint, warn
+        // messages should be recorded.
         if (lastCheckpointId < checkpointId) {
-            throw new IllegalStateException(
-                    String.format(
-                            "Gateway closed for different checkpoint: closed for = %d, expected = %d",
-                            currentCheckpointId, checkpointId));
+            LOG.warn("Trying to open gateway for unknown checkpoint: " + checkpointId);
```

Comment:
In the following case, we would have `lastCheckpointId < checkpointId`.

1. Suppose the coordinator has completed the checkpoint whose id is 10, and checkpoint barriers are sent out to snapshot the operators.
2. A subtask of the coordinator completed the snapshot, and send back an `AcknowledgeCloseGatewayEvent(checkpointId = 10)` (but the coordinator has not received the event yet).
3. For some reason, checkpoint 10 is discarded and the job is reset to checkpoint 5.
4. The coordinator receives the `AcknowledgeCloseGatewayEvent(checkpointId = 10)`.

This case happens because we introduced some operator events in the relevant PRs, and we cannot guarantee that the events can arrive at their destinations before the corresponding checkpoint finishes.